test/e2e/operator_test.go

```diff
@@ -62,240 +49,185 @@ func init() {
-		fmt.Printf("failed to get infrastructure config: %v\n", err)
-		os.Exit(1)
+func initProviderHelper() (providerTestHelper, error) {
+	var (
+		openshiftCI  bool
+		platformType string
+		err          error
+	)
+
+	if os.Getenv("OPENSHIFT_CI") != "" {
+		openshiftCI = true
+		platformType, err = getPlatformType(kubeClient)
+		if err != nil {
+			return nil, fmt.Errorf("failed to determine platform type: %w", err)
+		}
+	} else {
+		platformType = mustGetEnv("CLOUD_PROVIDER")
+	}
+
+	switch platformType {
+	case string(configv1.AWSPlatformType):
+		var (
+			awsAccessKeyID     string
+			awsSecretAccessKey string
+		)
+		if openshiftCI {
+			awsAccessKeyID, awsSecretAccessKey, err = rootAWSCredentials(kubeClient)
+			if err != nil {
+				return nil, fmt.Errorf("failed to get AWS credentials from CCO: %w", err)
+			}
+		} else {
+			awsAccessKeyID = mustGetEnv("AWS_ACCESS_KEY_ID")
+			awsSecretAccessKey = mustGetEnv("AWS_SECRET_ACCESS_KEY")
+		}
+		return newAWSHelper(awsAccessKeyID, awsSecretAccessKey)
+	default:
+		return nil, fmt.Errorf("unsupported Provider: '%s'", platformType)
 	}
+}
 
-	platform := infraConfig.Status.PlatformStatus
-	if platform == nil {
-		fmt.Printf("platform status is missing for infrastructure %s", infraConfig.Name)
+func TestMain(m *testing.M) {
+	var (
+		err error
+	)
+	if err = initKubeClient(); err != nil {
+		fmt.Print(err)
 		os.Exit(1)
 	}
-	switch platform.Type {
-	case configv1.AWSPlatformType:
-		mySession := awsSession.Must(awsSession.NewSession())
-		dnsClient.aws = route53.New(mySession)
-	default:
-		fmt.Printf("Unsupported Provider")
+
+	if helper, err = initProviderHelper(); err != nil {
+		fmt.Print(err)
 		os.Exit(1)
 	}
 
-	/*
-		defaultParentDomainName = os.Getenv("EXTDNS_PARENT_DOMAIN")
-		if defaultParentDomainName == "" {
-			fmt.Printf("No parent domain specified. Please set the environment variable EXTDNS_PARENT_DOMAIN\n")
-			os.Exit(1)
-		}
-		defaultParentZoneID = os.Getenv("EXTDNS_PARENT_ZONEID")
-		if defaultParentZoneID == "" {
-			fmt.Printf("No parent zone ID specified. Please set the environment variable EXTDNS_PARENT_ZONEID\n")
-			os.Exit(1)
-		}
-
-		domainName = generateDomainName()
+	hostedZoneID, nameServers, err = helper.ensureHostedZone(hostedZoneDomain)
+	if err != nil {
+		fmt.Printf("Failed to created hosted zone for domain %s: %v", hostedZoneDomain, err)
+		os.Exit(1)
+	}
 
-		dialer = &net.Dialer{
-			Resolver: &net.Resolver{
-				PreferGo: true,
-				Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
-					d := net.Dialer{
-						Timeout: dnsLookupTimeout,
-					}
-					return d.DialContext(ctx, "udp", dnsServer)
-				},
-			},
-		}
+	exitStatus := m.Run()
 
-		httpClient = &http.Client{
-			Transport: &http.Transport{
-				Dial:        dialer.Dial,
-				DialContext: dialer.DialContext,
-			},
-		}
-
-		os.Exit(m.Run())
-	*/
+	err = helper.deleteHostedZone(hostedZoneID)
```

Comment:
Do we want to `defer` this deletion right after successful `ensureHostedZone`?

Reply:
We cannot deferred calls because we have to call `os.Exit()` and that immediately exits the program. We can move all the logic into another function just to use defer but I think that might be overkill for now.

Reply:
Yes, it's kind of a choice between:
- last step cleanup (current implementation) which fails the test if the cleanup broke but leave leftover zones of the test cases panic
- best effort cleanup (no `Exit` with `defer`) which won't fail the test if the clean broke but will try to do something the test cases panic